src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs

```diff
@@ -0,0 +1,398 @@
+                                    [``public``; ``override``]
+                                    ``get`` (``=>`` body)
+            ``class`` "DumpRegister" ``<<`` ["T"] ``>>`` ``:`` 
+                (Some (simpleBase ("Quantum.Diagnostics.DumpRegister<T>"))) ``,`` [] [``public``] 
+                ``(`` [
+                    buildSimField ()
+                    buildConstructor "DumpRegister"
+                    dumpRegisterBody
+                ] ``)``
+        let qubitClasses = if generateQubitMgmt
+                           then
+                                seq { 
+                                    buildAllocClass "Allocate"
+                                    buildReturnClass "Release"
+                                    buildAllocClass "Borrow"
+                                    buildReturnClass "Return"
+                                }
+                           else
+                                Seq.empty
+        let dumpClasses = if generateDump 
+                          then seq { dumpMachineClass; dumpRegisterClass }
+                          else Seq.empty
+        Seq.append qubitClasses dumpClasses
+
+    let implMethodName (sp : QsSpecialization) =
+        "Do" + sp.Parent.Name.Value + match sp.Kind with
+                                      | QsBody              -> ""
+                                      | QsAdjoint           -> "Adj"
+                                      | QsControlled        -> "Ctl"
+                                      | QsControlledAdjoint -> "CtlAdj"
+
+    let fullArgsForSpecialization args (sp : QsSpecialization) =
+        match sp.Kind with
+        | QsControlled | QsControlledAdjoint -> ("__ctrls__", "IQArray<Qubit>") :: args
+        | _ -> args
+
+    let buildImplMethod context args (sp : QsSpecialization) =
+        let outType = match sp.Signature.ReturnType.Resolution with
+                      | UnitType -> "void"
+                      | _ -> SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        match sp.Implementation with
+        | Generated SelfInverse -> Seq.empty
+        | _ ->
+            let implName = implMethodName sp
+            let fullArgs = fullArgsForSpecialization args sp
+            let parms = fullArgs |> List.map (fun (n, t) -> ``param`` n ``of`` (``type`` t))
+            let impl = ``method`` outType implName ``<<`` [] ``>>`` 
+                            ``(`` parms ``)`` [``public``;``abstract``]
+                            ``{`` [] ``}`` :> MemberDeclarationSyntax
+            seq { impl }
+
+    let buildImplProperty targetClass context args (sp : QsSpecialization) =
+        let inType  = SimulationCode.roslynTypeName context sp.Signature.ArgumentType
+        let outType = SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        let propertyType = "Func<" + inType + ", " + outType + ">"
+        let propertyName =
```

Comment:
Since these need to match the generated C# code, it would be good to split out the corresponding functionality in the C# generation and invoke that one instead.

Reply:
Done.